heron/api/src/java/com/twitter/heron/streamlet/impl/groupings/JoinCustomGrouping.java

```diff
@@ -47,6 +47,9 @@ public void prepare(TopologyContext context, String component,
     List<Integer> ret = new ArrayList<>();
     V obj = (V) values.get(0);
     int index = keyExtractor.apply(obj).hashCode() % taskIds.size();
+    if (index < 0) {
+      index = (index + taskIds.size()) % taskIds.size();
```

Comment:
why are we doing this?

Reply:
how can index but less than 0?

Reply:
hashCode can return negative numbers. % according to spec can be negative if dividend is negative(it really is implementation specific, but looks like atleast some versions of jvm return negative).